### ISSUE TITLE
 Webview Process Terminated when instantiating shared workers on IOS 16.1

### ISSUE THREAD

**Issue by mauri870** [react-native-webview/react-native-webview#2718]
**Bug description:**

Some of our users are facing crashes related to the webview when opening our app on IOS 16.1 devices. The only message we could get from the crash is "Webview Process terminated".

**To Reproduce:**

Clone this [POC repository](https://github.com/mauri870/react-native-webview-ios-16.1-crash) and follow the required steps, it should crash when testing the react-native-webview on ios 16.1 but should work fine with 16.0

We already have a check to make sure we only use shared workers if they are available (typeof window.SharedWorker === 'function') but it crashes when doing `new SharedWorker`.

**Expected behavior:**

We should get a javascript exception/error instead of the webview process crashing.

**Screenshots/Videos:**

**Environment:**
 - OS: IOS
 - OS version: 16.1
 - react-native version: 0.68.2
 - react-native-webview: 11.23.1

**Comment by pparidans** [comment]
Hi!
We encounter a (very) similar issue with one of our mobile app since the release of iOS 16.1, but, in our case with a barebone `WKWebView` (no react-native involded). Not sure yet if there is a bug or a change introduced with this new iOS version...

@mauri870 Watch out that your POC uses the deprecated `UIWebView` instead of `WKWebView`.

Note: Safari on iOS 16.1 doesn't seam to be impacted.

**Comment by mauri870** [comment]
Alright, some updates.

Seems like UIWebView is deprecated, I updated the poc to use WKWebView which seems like the same api that this project uses and now it crashes too when instantiating shared workers, this was introduced by IOS 16.1 because it works fine with IOS 16.0.

I know that shared workers are not quite supported by the webkit webview but that crashing the app all of a sudden after an update does not seems right, also there is no mentioned changes in the 16.1 changelog.

**Comment by mauri870** [comment]
@pparidans Thanks for the feedback, I was already working on the migration to WKWebView. :)

To share our solution for now, seems like you cannot rely on just checking `typeof window.SharedWorker === 'function'` to see if the api is available, you need to blacklist 16.1 somehow, since we were already using detector-js we decided to simply disable shared workers and fallback to websockets (we only use sw for pusher events).

**Comment by pparidans** [comment]
@mauri870 Yes, I agree that even if it affects react-native-webview, it's not their responsibility (but Apple's one).
I also dug into the release notes or what else and found nothing! I guess we could treat it as a `WKWebView` bug then...

You mentioned that `SharedWorker` are not quite supported by the WKWebView : do you have a source for that (try to document this issue) ?

About the workaround, we also blacklisted it for lack of a better solution...

Anyway, thanks for your feedback :hugs:

**Comment by mauri870** [comment]
@pparidans Ended up creating an [Apple Developers forum post](https://developer.apple.com/forums/thread/718757).

This does not seems to be a Webkit issue at first, since it is the same version as in safari mobile.

I saw this ticket earlier about service workers, not sure if it is related https://github.com/react-native-webview/react-native-webview/issues/1086

Safari started supporting shared workers early this year, seems like it is not quite stable on mobile yet 😓

**Comment by mauri870** [comment]
Turns out this is a webkit bug on iOS 16.1

https://github.com/WebKit/WebKit/pull/5872

The fix for now is to disable shared workers by  using `injectedJavaScriptBeforeContentLoaded` with  `delete window.SharedWorker`

**Comment by LazyBoxer** [comment]
I have tried `injectedJavaScriptBeforeContentLoaded` with `delete window.SharedWorker`
But the problem still there. (my case: `source={{ html: ... }}`)

Then I found the post and tried to inject `delete window.SharedWorker` in native side.
https://developer.apple.com/forums/thread/718757
It works!

**Comment by ddiazFG** [comment]
> 
 Based on https://developer.apple.com/forums/thread/718757

Try:

```objective-c
#if defined(__IPHONE_OS_VERSION_MAX_ALLOWED) && __IPHONE_OS_VERSION_MAX_ALLOWED >= __IPHONE_16_1 /* iOS 16.1 */
      if (@available(iOS 16.1, *)) {
          NSString *deleteSharedWorker = @"delete window.SharedWorker; window.SharedWorker = undefined";
          WKUserScript *deleteSharedWorkerScript = [[WKUserScript alloc] initWithSource:deleteSharedWorker injectionTime:WKUserScriptInjectionTimeAtDocumentStart forMainFrameOnly:NO];
          [wkWebViewConfig.userContentController addUserScript:deleteSharedWorkerScript];
      }
    #endif
```

before html5HistoryAPIShimSource:

`[wkWebViewConfig.userContentController addUserScript:script];`

on RNCWebView

This is an IOS webkit issue so i don’t recommend patch this.
